server/src/config_loader.rs

```diff
@@ -0,0 +1,163 @@
+        Ok(config.apply_rocksdb_common())
+    }
+
+    fn merge_with_env(figment: Figment) -> Figment {
+        figment
+            .merge(Env::prefixed("RESTATE_").split("__"))
+            // Override tracing.log with RUST_LOG, if present
+            .merge(Env::raw().only(&["RUST_LOG"]).map(|_| "log_filter".into()))
+            .merge(
+                Env::raw()
+                    .only(&["HTTP_PROXY"])
+                    .map(|_| "http_proxy".into()),
+            )
+            .merge(
+                Env::raw()
+                    .only(&["AWS_EXTERNAL_ID"])
+                    .map(|_| "aws_assume_role_external_id".into()),
+            )
+            .merge(
+                Env::raw()
+                    .only(&["AWS_EXTERNAL_ID"])
+                    .map(|_| "aws_assume_role_external_id".into()),
+            )
+    }
+
+    pub fn start(self) {
+        if self.disable_watch || self.path.is_none() {
+            return;
+        }
+
+        let path = self.path.clone().unwrap();
+        let (tx, rx) = std::sync::mpsc::channel();
+        // Automatically select the best implementation for watching files on
+        // the current platform.
+        let Ok(mut debouncer) = new_debouncer(
+            Duration::from_secs(3),
+            move |res: DebounceEventResult| match res {
+                Ok(events) => tx.send(events).unwrap(),
+                Err(e) => warn!("Error {:?}", e),
+            },
+        ) else {
+            warn!(
+                "Couldn't initialize configuration watcher, config changes will not be monitored",
+            );
+            return;
+        };
+
+        info!("Installing watcher for config changes: {}", path.display());
+        if let Err(e) = debouncer
+            .watcher()
+            .watch(&path, notify::RecursiveMode::NonRecursive)
+        {
+            warn!("Couldn't install configuration watcher: {}", e);
+            return;
+        };
+
+        std::thread::Builder::new()
+            .name("config-watcher".to_owned())
+            .spawn(move || {
+                // It's important that we capture the watcher in the thread,
+                // otherwise it'll be dropped and we won't be watching anything!
+                let _debouncer = debouncer;
+                info!("Configuration watcher thread has started");
+                let mut should_run = true;
+                while should_run {
+                    match rx.recv() {
+                        Ok(evs) => {
+                            self.handle_events(evs);
+                        }
+                        Err(e) => {
+                            error!("Cannot continue watching configuration changes: '{}!", e);
+                            should_run = false;
+                        }
+                    }
+                }
+                info!("Config watcher thread has terminated");
+            })
+            .expect("start config watcher thread");
```

Comment:
What is the main purpose of this additional thread? Could we move `self.handle_events` into the `DebounceEventResult` handler?